### PR TITLE
wgsl: Replace FPTraits with static data for AF addition execution tests

### PR DESCRIPTION
**Pull request by zoddicus**
This also introduces the reference C++ program, af_data_gen, for how these values are generated. This code is not meant to be used as part of the CTS, so does not have any build system included with it.

Issue #2993

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [x] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Review by dj2 — APPROVED** [review]

File: tools/af_data_gen/main.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+std::string printBinaryCase(const std::string &lhs, const std::string &rhs, const std::vector<double> &result) {
+    assert(!result.empty());
+    std::stringstream ss;
+    ss << "{ lhs: ";
+    ss << lhs;
+    ss << ", rhs: ";
+    ss << rhs;
+    ss << ", ";
+    ss << "expected: [ ";
+    for (auto i = 0; i < result.size() - 1; i++) {
```
[the remark this answers is not in the document]
Comment (dneto0) [reply tools/af_data_gen/main.cpp:123]: ...or use a flag.

File: tools/af_data_gen/main.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
+/**
+ * Code used to generate manual values for 'correctly rounded' AbstractFloat tests in the CTS.
+ *
+ * These are generated in a C++ program, because it allows for easy access to higher than 64-bit floating point numbers
```
Comment (dneto0) [inline tools/af_data_gen/main.cpp:4]: This is a pretty nice way to do it.

**Review by dneto0 — APPROVED** [review]
This is very nice.
thanks!

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+/** Line format af_data_gen/main.cpp outputs */
+interface RawBinaryData {
+  lhs: number;
+  rhs: number;
+  expected: number[];
+}
+
+// prettier-ignore
+export const addition_raw_values: RawBinaryData[] = [
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:15]: nit: please add a doc comment explaining what's in here

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+  return crc32(`lhs: ${lhs}, rhs: ${rhs}`);
+}
+
+/** Table mapping addition binary inputs to expectation intervals */
+const addition_table: Record<number, FPInterval> = {};
+addition_raw_values.forEach(value => {
+  addition_table[hashRawDataInputs(value.lhs, value.rhs)] = FP.abstract.spanIntervals(
+    ...value.expected.map(e => FP.abstract.correctlyRoundedInterval(e))
+  );
+});
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:255]: nit: Prefer `for..of`, but better, construct all at once so the result can be readonly:

```suggestion
const kAdditionTable: Readonly<Record<number, FPInterval>> = Object.fromEntries(
  kAdditionRawValues.map(value => [
    hashRawDataInputs(value.lhs, value.rhs),
    FP.abstract.spanIntervals(...value.expected.map(e => FP.abstract.correctlyRoundedInterval(e))),
  ])
);
```

(this should work, but please make sure)

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+import { kValue } from '../../../../util/constants.js';
+import { FP, FPInterval } from '../../../../util/floating_point.js';
+import { sparseScalarF64Range } from '../../../../util/math.js';
+import { reinterpretU64AsF64 } from '../../../../util/reinterpret.js';
+
+/** Line format af_data_gen/main.cpp outputs */
+interface RawBinaryData {
+  lhs: number;
+  rhs: number;
+  expected: number[];
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:11]: Should add a few `readonly`s and `as const`s to make sure there's no mutable global state in this file. Suggestions here and below.

```suggestion
  readonly lhs: number;
  readonly rhs: number;
  readonly expected: readonly number[];
```

File: src/webgpu/shader/execution/expression/binary/af_data.ts (hunk 3 above)
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:15]:
```suggestion
export const addition_raw_values: readonly RawBinaryData[] = [
```

Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:15]: nit: rename to kAdditionRawValues

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+];
+
+/** Hashing function for keys in lookup tables */
+function hashRawDataInputs(lhs: number, rhs: number): number {
+  return crc32(`lhs: ${lhs}, rhs: ${rhs}`);
+}
+
+/** Table mapping addition binary inputs to expectation intervals */
+const addition_table: Record<number, FPInterval> = {};
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:250]: nit: rename to kAdditionTable

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+export function getAdditionAFInterval(lhs: number, rhs: number): FPInterval {
+  return addition_table[hashRawDataInputs(lhs, rhs)];
+}
+
+/** Simplified version of kSparseVectorF64Range that only uses values from sparseScalarF64Range() */
+export const kSparseVectorAFValues = {
+  2: sparseScalarF64Range().map(f => Array(2).fill(f)),
+  3: sparseScalarF64Range().map(f => Array(3).fill(f)),
+  4: sparseScalarF64Range().map(f => Array(4).fill(f)),
+};
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:267]:
```suggestion
} as const;
```

File: src/webgpu/shader/execution/expression/binary/af_data.ts [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    2: sparseScalarF64Range().map(f => Array(3).fill(Array(2).fill(f))),
+    3: sparseScalarF64Range().map(f => Array(3).fill(Array(3).fill(f))),
+    4: sparseScalarF64Range().map(f => Array(3).fill(Array(4).fill(f))),
+  },
+  4: {
+    2: sparseScalarF64Range().map(f => Array(4).fill(Array(2).fill(f))),
+    3: sparseScalarF64Range().map(f => Array(4).fill(Array(3).fill(f))),
+    4: sparseScalarF64Range().map(f => Array(4).fill(Array(4).fill(f))),
+  },
+};
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/binary/af_data.ts:286]:
```suggestion
} as const;
```

**Review by kainino0x — COMMENTED** [review]
Deferring review of the meat of this PR to others, but structurally LGTM. A few small comments.

File: tools/af_data_gen/main.cpp [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    return ss.str();
+}
+
+int main() {
+    assert(sizeof(double) < sizeof(long double) && "Need higher precision long double");
+    assert(sizeof(long double) == 16 && "Code assume 'proper' quad support, not some other higher precision floating point implementation");
+
+    {
+        // Confirm that calculating f64 imprecise results generates two possible roundings.
+        const auto [begin, end] = quantizeToAbstractFloat(static_cast<long double>(0.1) * static_cast<long double>(0.1));
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply tools/af_data_gen/main.cpp:183]: Thanks, the number of caveats that I was seeing about these controls may not be implemented in the standard lib or the long double format may not be quad if the CPU doesn't implementat quad instructions, behooved me to add a check that the results returned are what is expected, so if someone needs to run this in the future on a different machine in the future.

**Pull request merged**